# Assignment: Data Wrangling
### `! git clone https://github.com/ds3001f25/wrangling_assignment.git`
### Do Q1 and Q2
### Reading material: `tidy_data.pdf`

**Q1.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? 
3. Dummy variable: For the pretrial data covered in the lecture `./data/justice_data.parquet`, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [1]:
%pip install pandas
%pip install numpy
%pip install xlrd
%pip install re

import pandas as pd
import numpy as np

#Q1.1
df_price = pd.read_csv("./data/airbnb_hw.csv", low_memory=False)

df_price['Price'] = df_price['Price'].astype(str) # str type to be able to remove commas
df_price['Price'] = df_price['Price'].str.strip().replace(',','') # removing commas
df_price['Price'] = pd.to_numeric(df_price['Price'], errors='coerce') # converting type into float


df_price['Price_NA'] = df_price['Price'].isnull() # check for null values
print(df_price.head())
print('Total missing values:', sum(df_price['Price_NA']))


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: C:\Users\dsong\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: C:\Users\dsong\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: C:\Users\dsong\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip
ERROR: Could not find a version that satisfies the requirement re (from versions: none)

[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: C:\Users\dsong\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip
ERROR: No matching distribution found for re


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.
    Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0  

2. Categorical variable: For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? 

In [2]:
#Q1.2
df_injury = pd.read_csv("./data/mn_police_use_of_force.csv", low_memory=False)

missing = df_injury['subject_injury'].isnull()
print('Proportion missing:', missing.mean()*100, "%")
print("Cleaned subject_injury:", df_injury['subject_injury'].unique())
df_injury['subject_injury'] = df_injury['subject_injury'].replace(np.nan, 'Missing')
df_injury['subject_injury'] = df_injury['subject_injury'].str.strip()

ctab = pd.crosstab(df_injury['subject_injury'], df_injury['force_type'])
print("Cross-tabulation:\n", ctab.head())

Proportion missing: 76.19342359767892 %
Cleaned subject_injury: [nan 'No' 'Yes']
Cross-tabulation:
 force_type      Baton  Bodily Force  Chemical Irritant  Firearm  \
subject_injury                                                    
Missing             2          7051               1421        0   
No                  0          1093                131        2   
Yes                 2          1286                 41        0   

force_type      Gun Point Display  Improvised Weapon  Less Lethal  \
subject_injury                                                      
Missing                        27                 74           87   
No                             33                 34            0   
Yes                            44                 40            0   

force_type      Less Lethal Projectile  Maximal Restraint Technique  \
subject_injury                                                        
Missing                              0                          170   
No   

About 76% of the 'subject_injury' values are missing, which is quite high and definitely a concern.
The cross tabulation shows that the majority of missing values for 'subject_injury' come from common force types such as Bodily Force and Chemical Irritant.

3. Dummy variable: For the pretrial data covered in the lecture ./data/justice_data.parquet, clean the WhetherDefendantWasReleasedPretrial variable as well as you can, and, in particular, replace missing values with np.nan. 4. Missing values, not at random: For the pretrial data covered in the lecture, clean the ImposedSentenceAllChargeInContactEvent variable as well as you can, and explain the choices you make. (Hint: Look at the SentenceTypeAllChargesAtConvictionInContactEvent variable.)

In [3]:
#Q1.3
df_justice = pd.read_parquet("./data/justice_data.parquet", engine="pyarrow")
print("Before cleaning:", df_justice["WhetherDefendantWasReleasedPretrial"].unique())
df_justice['WhetherDefendantWasReleasedPretrial'] = df_justice['WhetherDefendantWasReleasedPretrial'].replace(9,np.nan)
print("After cleaning:", df_justice["WhetherDefendantWasReleasedPretrial"].unique())

Before cleaning: [9 0 1]
After cleaning: [nan  0.  1.]


4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [4]:
#Q1.4

print("Before cleaning:", df_justice['ImposedSentenceAllChargeInContactEvent'].head())
df_justice['ImposedSentenceAllChargeInContactEvent'] = df_justice['ImposedSentenceAllChargeInContactEvent'].replace(' ', np.nan) # replacing with np.nan

# Step 2: Convert to numeric (some were strings like '12.98...')
df_justice['ImposedSentenceAllChargeInContactEvent'] = pd.to_numeric(df_justice['ImposedSentenceAllChargeInContactEvent'], errors='coerce' # convert
)

# Step 3: Replace Sentence Type code 9 ("unknown") with NaN
df_justice['SentenceTypeAllChargesAtConvictionInContactEvent'] = df_justice['SentenceTypeAllChargesAtConvictionInContactEvent'].replace(9, np.nan) # replace 9 with NaN

# Step 4: If SentenceType == 0 (likely "no sentence"), then set sentence value to NaN
df_justice.loc[df_justice['SentenceTypeAllChargesAtConvictionInContactEvent'] == 0,
    'ImposedSentenceAllChargeInContactEvent'] = np.nan

print("After cleaning:", df_justice['ImposedSentenceAllChargeInContactEvent'].head())

print("Missing values:", df_justice['ImposedSentenceAllChargeInContactEvent'].isna().sum())
print("Total rows:", len(df_justice))

print("Unique sentence types:")
print(df_justice['SentenceTypeAllChargesAtConvictionInContactEvent'].dropna().unique())



Before cleaning: 0                    
1                  60
2                  12
3    .985626283367556
4                    
Name: ImposedSentenceAllChargeInContactEvent, dtype: object
After cleaning: 0          NaN
1          NaN
2    12.000000
3     0.985626
4          NaN
Name: ImposedSentenceAllChargeInContactEvent, dtype: float64
Missing values: 17773
Total rows: 22986
Unique sentence types:
[0. 1. 4. 2.]


**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks (Hint: `GSAF5.xls`).

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?
7. Clean the `Fatal Y/N` variable so it only takes three values: Y, N, and Unknown.
8. Are sharks more likely to launch unprovoked attacks on men or women? Is the attack more or less likely to be fatal when the attack is provoked or unprovoked? Is it more or less likely to be fatal when the victim is male or female? How do you feel about sharks?
9. What proportion of attacks appear to be by white sharks? (Hint: `str.split()` makes a vector of text values into a list of lists, split by spaces.)

In [5]:
import re
# Q2.1
df = pd.read_excel("GSAF5.xls")
df.head()


Date    Year        Type      Country  \
0  16th August 2025  2025.0    Provoked          USA   
1       18th August  2025.0  Unprovoked    Australia   
2       17th August  2025.0  Unprovoked      Bahamas   
3        7th August  2025.0  Unprovoked    Australia   
4        1st August  2025.0  Unprovoked  Puerto Rico   

                               State                             Location  \
0                            Florida               Cayo Costa Boca Grande   
1                                NSW                       Cabarita Beach   
2  Atlantic Ocean near Big Grand Cay  North of Grand Bahama near Freeport   
3                                NSW                        Tathra Beach    
4                           Carolina                       Carolina Beach   

       Activity          Name Sex Age  ...                 Species   \
0       Fishing   Shawn Meuse   M   ?  ...  Lemon shark 1.8 m (6ft)   
1       Surfing     Brad Ross   M   ?  ...  5m (16.5ft) Great White   
2  Spearfishing    Not stated   M  63  ...             Undetermined   
3       Surfing   Bowie Daley   M   9  ...    Suspected Great White   
4        Wading  Eleonora Boi   F  39  ...             Undetermined   

                                              Source  pdf href formula href  \
0  Johannes Marchand: Kevin McMurray Trackingshar...  NaN          NaN  NaN   
1  Bob Myatt GSAF The Guardian: 9 News: ABS News:...  NaN          NaN  NaN   
2  Ralph Collier GSAF and Kevin MCMurray Tracking...  NaN          NaN  NaN   
3                                     Bob Myatt GSAF  NaN          NaN  NaN   
4         Kevin McMurray Trackingsharks.com: NY Post  NaN          NaN  NaN   

  Case Number Case Number.1 original order Unnamed: 21 Unnamed: 22  
0         NaN           NaN            NaN         NaN         NaN  
1         NaN           NaN            NaN         NaN         NaN  
2         NaN           NaN            NaN         NaN         NaN  
3         NaN           NaN            NaN         NaN         NaN  
4         NaN           NaN            NaN         NaN         NaN  

[5 rows x 23 columns]

In [7]:

# Q2.2
df = df.dropna(axis=1, how='all')  # drop empty columns
print("Columns:", df.columns.tolist())


Columns: ['Date', 'Year', 'Type', 'Country', 'State', 'Location', 'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal Y/N', 'Time', 'Species ', 'Source', 'pdf', 'href formula', 'href', 'Case Number', 'Case Number.1', 'original order', 'Unnamed: 21', 'Unnamed: 22']


In [8]:

# Q2.3
df['Year'] = pd.to_numeric(df['Year'], errors='coerce')  # numeric
df = df[df['Year'].between(1940, 2025)]  # year change
print("Year range:", df['Year'].min(), ",", df['Year'].max())

attacks_per_year = df["Year"].value_counts().sort_index() # look at trend over time
print("First few year counts:\n", attacks_per_year.head())
print("Last few year counts:\n", attacks_per_year.tail())

early_mean = attacks_per_year.loc[1900:1950].mean() # filtering based on past attacks
recent_mean = attacks_per_year.loc[2000:].mean() # filtering based on recent attacks
if recent_mean > early_mean * 1.25: # recent rate must be at least 25% higher to be increasing
    trend = "increasing overall"
elif recent_mean < early_mean * 0.75: # recent rate must be at least 25% lower to be decreasing
    trend = "decreasing overall"
else:
    trend = "roughly constant"
print("Overall trend in shark attacks over time:", trend)


Year range: 1940.0 , 2025.0
First few year counts:
 Year
1940.0    24
1941.0    27
1942.0    41
1943.0    28
1944.0    31
Name: count, dtype: int64
Last few year counts:
 Year
2021.0    110
2022.0     98
2023.0    109
2024.0     52
2025.0     48
Name: count, dtype: int64
Overall trend in shark attacks over time: increasing overall


In [9]:

# Q2.4
df['Age'] = pd.to_numeric(df['Age'], errors='coerce')  # convert
print("Age summary:", df['Age'].describe())


Age summary: count    3599.000000
mean       28.600445
std        14.796239
min         1.000000
25%        17.000000
50%        25.000000
75%        37.000000
max        87.000000
Name: Age, dtype: float64


In [10]:

# Q2.5
sex_counts = df['Sex'].value_counts(dropna=False) # finding number of entries
sex_proportions = sex_counts / sex_counts.sum() # finding proportion
print("Proportions:", sex_proportions)


Proportions: Sex
M        0.786052
F        0.130222
NaN      0.082092
M        0.000545
F        0.000363
 M       0.000182
m        0.000182
lli      0.000182
M x 2    0.000182
Name: count, dtype: float64


In [11]:

# Q2.6
df['Type'] = df['Type'].astype(str)
df['Type'] = df['Type'].str.lower()
df['Type'] = df['Type'].where(df['Type'].isin(['provoked', 'unprovoked']), 'unknown') # using provoked and unprovoked, everything else goes into unknown

type_counts = df['Type'].value_counts(dropna=False)
print("Attack types:", type_counts)
print("Unprovoked proportion:", type_counts['unprovoked'] / type_counts.sum())




Attack types: Type
unprovoked    4097
unknown        891
provoked       518
Name: count, dtype: int64
Unprovoked proportion: 0.7440973483472575


In [12]:
# Q2.7
df['Fatal Y/N'] = df['Fatal Y/N'].astype(str)
df['Fatal Y/N'] = df['Fatal Y/N'].str.lower()
df['Fatal Y/N'] = df['Fatal Y/N'].replace({
    "y": "Y",
    "yes": "Y",
    "y x 2": "Y",
    "n": "N",
    "no": "N",
    "f": "N",
})
df['Fatal Y/N'] = np.where(df['Fatal Y/N'].isin(["Y","N"]), df['Fatal Y/N'], "Unknown")

print("Value Counts:", df['Fatal Y/N'].value_counts())
print("Unique values:", df['Fatal Y/N'].unique())




Value Counts: Fatal Y/N
N          4218
Y           825
Unknown     463
Name: count, dtype: int64
Unique values: ['N' 'Y' 'Unknown']


In [13]:
# Q2.8
print("Attacks by Sex and Type:\n", 
      pd.crosstab(df['Sex'], df['Type'], normalize='index').to_string())
print("Fatality by Type:\n", 
      pd.crosstab(df['Type'], df['Fatal Y/N'], normalize='index').to_string())

print("\nFatality by Sex:\n", 
      pd.crosstab(df['Sex'], df['Fatal Y/N'], normalize='index').to_string())






Attacks by Sex and Type:
 Type   provoked   unknown  unprovoked
Sex                                  
 M     0.000000  0.000000    1.000000
F      0.039052  0.110181    0.850767
F      0.000000  0.000000    1.000000
M      0.103050  0.125693    0.771257
M      0.000000  0.000000    1.000000
M x 2  0.000000  1.000000    0.000000
lli    0.000000  0.000000    1.000000
m      0.000000  0.000000    1.000000
Fatality by Type:
 Fatal Y/N          N   Unknown         Y
Type                                    
provoked    0.955598  0.021236  0.023166
unknown     0.414141  0.445567  0.140292
unprovoked  0.818648  0.013424  0.167928

Fatality by Sex:
 Fatal Y/N         N   Unknown         Y
Sex                                    
 M         1.000000  0.000000  0.000000
F          0.792190  0.083682  0.124128
F          1.000000  0.000000  0.000000
M          0.773105  0.074861  0.152033
M          0.666667  0.000000  0.333333
M x 2      0.000000  0.000000  1.000000
lli        1.000000  0.000000  

- The table shows that about 85% of attacks on women are unprovoked, whereas about 77% of attacks on men are unprovoked. Therefore, unprovoked attacks are more likely to happen to women than men.

- The table shows that provoked attacks are only fatal about 2.3% of the time and unprovoked attacks are fatal about 16.8% of the time. Therefore, unprovoked attacks are more likely to be fatal than provoked attacks.

- The table shows that of female victims, 12.4% of the attacks are fatal, whereas for male victims, 15.2% of attacks are fatal. Therefore, attacks on men are more likely to be fatal than those on women. 

- Although shark attacks are rare, the data shows that they are usually unprovoked and more dangerous when they are. Men are more frequently attacked, but women’s attacks are even more likely to be unprovoked. The results suggest that sharks are not malicious, but act based on their instincts. 



In [14]:
# Q2.9
df['Species '] = df['Species '].astype(str)
df['Species '] = df['Species '].str.lower().str.strip() # clean

species_split = df['Species '].str.split() # split by spaces into lists

is_white = species_split.apply(lambda x: ('white' in x)) # find rows with white

count_white = is_white.sum()
total = df['Species '].notna().sum()
print("Rows with white:", count_white, "out of", len(df), "total")
print("Proportion of attacks from white sharks: ", count_white/len(df))


Rows with white: 668 out of 5506 total
Proportion of attacks from white sharks:  0.1213221939702143
